peekingduck/pipeline/nodes/model/jde_mot/jde_files/multitracker.py

```diff
@@ -0,0 +1,546 @@
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+Script for creating Track and JDE model instance
+"""
+
+from typing import Any, Dict, List, Tuple
+from collections import deque
+import numpy as np
+import torch
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.models import (
+    Darknet,
+)
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files import matching
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.basetrack import (
+    BaseTrack,
+    TrackState,
+)
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.utils.kalman_filter import (
+    KalmanFilter,
+)
+from peekingduck.pipeline.nodes.model.jde_mot.jde_files.utils.utils import (
+    non_max_suppression,
+    scale_coords,
+)
+
+# mypy: ignore-errors
+# pylint: disable=invalid-name, too-many-instance-attributes, too-many-locals, too-many-branches, too-many-statements, too-few-public-methods, arguments-differ, no-member, pointless-string-statement, useless-object-inheritance
+
+
+class STrack(BaseTrack):
+    """STrack class"""
+
+    def __init__(
+        self,
+        tlwh: torch.Tensor,
+        score: torch.Tensor,
+        temp_feat: torch.Tensor,
+        buffer_size: int = 30,
+    ) -> None:
+
+        # wait activate
+        self._tlwh = np.asarray(tlwh, dtype=np.float)
+        self.kalman_filter = None
+        self.mean, self.covariance = None, None
+        self.is_activated = False
+
+        self.score = score
+        self.tracklet_len = 0
+
+        self.smooth_feat = None
+        self.update_features(temp_feat)
+        self.features = deque([], maxlen=buffer_size)
+        self.alpha = 0.9
+
+    def update_features(self, feat: torch.Tensor) -> None:
+        """Update features.
+
+        Args:
+            feat (torch.Tensor): Features of predictions.
+        """
+        feat /= np.linalg.norm(feat)
+        self.curr_feat = feat
+        if self.smooth_feat is None:
+            self.smooth_feat = feat
+        else:
+            self.smooth_feat = self.alpha * self.smooth_feat + (1 - self.alpha) * feat
+        self.features.append(feat)
+        self.smooth_feat /= np.linalg.norm(self.smooth_feat)
+
+    def predict(self) -> None:
+        """Predict function."""
+        mean_state = self.mean.copy()
+        if self.state != TrackState.Tracked:
+            mean_state[7] = 0
+        self.mean, self.covariance = self.kalman_filter.predict(
+            mean_state, self.covariance
+        )
+
+    @staticmethod
+    def multi_predict(stracks: List[Any], kalman_filter: KalmanFilter) -> None:
```

Comment:
more detailed description of type of list content

Reply:
As per our discussion, type of list is STracks and this is mentioned in the description.